File: 30_oct_2019.cpp
/*
Question:
Given a list of numbers, where every number shows up twice except for one number, find that number

Time Complexity: O(n)
Space Complexity: O(1)
*/

#include <iostream>
#include <vector>

using namespace std;

class Solution {
public:
    int findUnique(vector<int> nums) {
        int p = 1;
        bool zero = false;

        for(auto c: nums) {
            if (c == 0) {
                zero = !zero;
                continue;
            }
            if (p % c ==0) {
                p /= c;
            }
            else {
                p *= c;
            }
        }

        if (zero) {
            return 0;
        }

        return p;
    }
};

int main() {
    Solution s;
    int a = s.findUnique(vector<int>{0, 2, 1, 1, 2, 3, 4, 4, 3});
    cout << a << endl;
}

File: 30_oct_2019.py
"""
Question:
Given a list of numbers, where every number shows up twice except for one number, find that number

Idea:
Credits - @Pratik.Somwanshi
> Take XOR of all the numbers. Same numbers will cancel out each other.

Time Complexity: O(n)
Space Complexity: O(1)
"""
import unittest
from functools import reduce

class Solution:

	def findUnique(self, nums):
		return reduce(lambda x, y: x ^ y, nums)

class TestCases(unittest.TestCase):

	def test_findUnique(self):
		s = Solution()
		self.assertEqual(s.findUnique([1, 2, 3, 2, 3]), 1)
		self.assertEqual(s.findUnique([2, 3, 2]), 3)
		self.assertEqual(s.findUnique([2]), 2)
		self.assertEqual(s.findUnique([0, 1, 2, 2, 1]), 0)
		self.assertEqual(s.findUnique([0, 0, 1, 1, 2]), 2)
		self.assertEqual(s.findUnique([0, 0, 1, 1, 2]), 2)


if __name__ == '__main__':
	s = Solution()